pkg/ingester/metadata_map.go

```diff
@@ -0,0 +1,153 @@
+package ingester
+
+import (
+	"sync"
+	"time"
+
+	"github.com/prometheus/common/model"
+
+	"github.com/cortexproject/cortex/pkg/ingester/client"
+)
+
+type metadataMap struct {
+	mtx              sync.Mutex
+	metricToMetadata map[string]*memoryMetadataSet
+}
+
+func newMetadataMap() *metadataMap {
+	return &metadataMap{
+		metricToMetadata: map[string]*memoryMetadataSet{},
+	}
+}
+
+func (mm *metadataMap) purge(max time.Duration) int {
+	mm.mtx.Lock()
+	defer mm.mtx.Unlock()
+	var deleted int
+	now := model.Now()
+	for m, s := range mm.metricToMetadata {
+		deleted += s.purge(now, max)
+
+		if s.length() <= 0 {
+			delete(mm.metricToMetadata, m)
+		}
+	}
+
+	return deleted
+}
+
+func (mm *metadataMap) all() map[string]*memoryMetadataSet {
+	mm.mtx.Lock()
+	defer mm.mtx.Unlock()
+	metadata := mm.metricToMetadata
+	return metadata
+}
+
+func (mm *metadataMap) get(metric string) (*memoryMetadataSet, bool) {
+	mm.mtx.Lock()
+	defer mm.mtx.Unlock()
+	set, ok := mm.metricToMetadata[metric]
+	if !ok {
+		return nil, false
+	}
+	return set, true
+}
+
+func (mm *metadataMap) put(metric string, set *memoryMetadataSet) {
+	mm.mtx.Lock()
+	defer mm.mtx.Unlock()
+	mm.metricToMetadata[metric] = set
+}
+
+func (mm *metadataMap) metadataLength() int {
+	var n int
+	mm.mtx.Lock()
+	for _, set := range mm.metricToMetadata {
+		n += set.length()
+	}
+	mm.mtx.Unlock()
+	return n
+}
+
+func (mm *metadataMap) length() int {
+	mm.mtx.Lock()
+	l := len(mm.metricToMetadata)
+	mm.mtx.Unlock()
+	return l
+}
+
+func newMemoryMetadataSet() *memoryMetadataSet {
+	return &memoryMetadataSet{
+		set: map[memoryMetadata]model.Time{},
+	}
+}
+
+type memoryMetadataSet struct {
+	mtx sync.Mutex
+	set map[memoryMetadata]model.Time
+}
+
+func (mms *memoryMetadataSet) all() map[memoryMetadata]model.Time {
+	mms.mtx.Lock()
+	defer mms.mtx.Unlock()
+	s := mms.set
+
+	return s
+}
+
+// add returns false if the item was already in the set or true if it wasn't.
+func (mms *memoryMetadataSet) add(m memoryMetadata) bool {
+	mms.mtx.Lock()
+	defer mms.mtx.Unlock()
+
+	_, ok := mms.set[m]
+	mms.set[m] = model.Now()
```

Comment:
[nit]: I would suggest to pass the time as parameter, and make all metadata with single push have same time.